Generate SwaggerUI

GuentherJulian opened the pull request:
Created a new template into which the OpenApi specification can be passed to create a completely working Swagger UI. Added a new parameter to specify the output directory of the files, so that the files can be served directly with an application (e.g. Quarkus, Spring) without the need for a separate web server.

GuentherJulian committed:
generate swagger ui

hohwille commented:
@GuentherJulian thanks for this PR. Looks nice with the swagger JS inclusion. 👍 
Can you resolve the conflicts?

GuentherJulian committed:
Merge remote-tracking branch 'upstream/master' into feature/generation_of_swaggerui

# Conflicts:
#	src/main/java/org/codehaus/mojo/servicedocgen/ServiceDocGenReport.java

GuentherJulian committed:
new parameter for generator

GuentherJulian commented:
> @GuentherJulian thanks for this PR. Looks nice with the swagger JS inclusion. 👍
> Can you resolve the conflicts?

Resolved them @hohwille

hohwille commented on src/main/java/org/codehaus/mojo/servicedocgen/generation/ServicesGenerator.java:
```diff
@@ -40,7 +40,7 @@
      * @throws IOException if something goes wrong.
      */
-    void generate( ServicesDescriptor descriptor, File outputDirectory, String filename )
+    void generate( ServicesDescriptor descriptor, File outputDirectory, String filenamet, String openApiUrl )
```
can you use refactoring (change method signature)?
You are breaking JavaDoc links. This again broke the site generation with the last PR and caused issues with the release.
Maybe we can even change the build so the JavaDoc is always generated and CI will fail in such case.

hohwille commented on src/main/java/org/codehaus/mojo/servicedocgen/ServiceDocGenReport.java:
```diff
@@ -162,6 +165,11 @@ public boolean isExternalReport()
         return true;
     }
+    
+    protected String getOutputDirectory()
```
Can you name it `getOutputDirectoryPath` instead to avoid violating JavaBeans conventions?
I am not always 100% sure how strict or lax the mavens internal introspection is but we might get issues if we do not strictly follow JavaBeans conventions.

hohwille commented on src/main/java/org/codehaus/mojo/servicedocgen/ServiceDocGenReport.java:
```diff
@@ -261,14 +270,30 @@ private void generateReport()
+                openApiUrl = "openapi.json";
+                break;
+            }
+        }
```
I have created a simple fallback for the output name that is simply omitting the `.vm` extension.
IMHO we should update this plugin to follow such simple conventions.
Why would we need this case differences `OpenApi.yaml.vm` vs. `openapi.yaml`?
Let us either rename the template to `openapi.yaml.vm` or change the default output to `OpenApi.yaml`.
We should also remove the default output names (see lines 179-181)...

hohwille reviewed (approved):
@GuentherJulian Thanks for this PR. It is a nice improvment. I have added some review comments for improvements. As they are more or less nice-to-have improvements, I already approve. If you happen to find some time to rework I would highly appreciate. Otherwise, let us know and I can merge and rework after merging...

GuentherJulian committed:
fixed javadoc, use fallback for output filenames

GuentherJulian commented:
> @GuentherJulian Thanks for this PR. It is a nice improvment. I have added some review comments for improvements. As they are more or less nice-to-have improvements, I already approve. If you happen to find some time to rework I would highly appreciate. Otherwise, let us know and I can merge and rework after merging...

@hohwille I have corrected your comments. If there is anything else to correct, let me know.
